agent/flux.go

```diff
@@ -1,36 +1,192 @@
+	// For requiring a particular registry to be accessible, else crash
+	RegistryRequire []string
+	// Can be used to switch image registry scanning off for some or
+	// all images (with glob patterns)
+	RegistryExcludeImage []string
+
+	// This is now hard-wired to empty in launch-generator, to tell flux
+	// _not_ to use service discovery. But: maybe someone needs to use
+	// service discovery.
+	MemcachedService *string
+
+	// Just in case we more explicitly support restricting Weave
+	// Cloud, or just Flux to particular namespaces
+	AllowNamespace []string
 }
 
-func getFluxConfig(k kubectl.Client, namespace string) (*FluxConfig, error) {
-	out, err := k.Execute("get", "pod", "-n", namespace, "-l", "name=weave-flux-agent", "-o", "jsonpath='{.items[?(@.metadata.labels.name==\"weave-flux-agent\")].spec.containers[0].args[*]}'")
-	if err != nil {
-		return nil, err
+// AsQueryParams returns the configuration as a fragment of query
+// string, so it can be interpolated into a text template.
+func (c *FluxConfig) AsQueryParams() string {
+	// Nothing clever here.
+	vals := url.Values{}
+
+	if c == nil {
+		return ""
+	}
+
+	// String-valued arguments
+	for arg, val := range map[string]*string{
+		"git-label":         c.GitLabel,
+		"git-url":           c.GitURL,
+		"git-branch":        c.GitBranch,
+		"memcached-service": c.MemcachedService,
+	} {
+		if val != nil {
+			vals.Add(arg, *val)
+		}
+	}
+
+	// []string-valued arguments
+	for arg, slice := range map[string][]string{
+		"git-path":               c.GitPath,
+		"registry-ecr-region":    c.RegistryECRRegion,
+		"registry-require":       c.RegistryRequire,
+		"registry-exclude-image": c.RegistryExcludeImage,
+		"k8s-allow-namespace":    c.AllowNamespace,
+	} {
+		for _, val := range slice {
+			vals.Add(arg, val)
+		}
+	}
+
+	// duration-valued arguments
+	for arg, dur := range map[string]*time.Duration{
+		"git-timeout":       c.GitTimeout,
+		"git-poll-interval": c.GitPollInterval,
+	} {
+		if dur != nil {
+			vals.Add(arg, dur.String())
+		}
 	}
 
+	for arg, flag := range map[string]*bool{
+		"sync-garbage-collection": c.GarbageCollection,
+		"git-set-author":          c.GitSetAuthor,
+		"git-ci-skip":             c.GitCISkip,
+	} {
+		if flag != nil {
+			vals.Add(arg, strconv.FormatBool(*flag))
+		}
+	}
+
+	return vals.Encode()
+}
+
+func isFlagPassed(fs *pflag.FlagSet, name string) bool {
+	found := false
+	fs.Visit(func(f *pflag.Flag) {
+		if f.Name == name {
+			found = true
+		}
+	})
+	return found
+}
+
+// ParseFluxArgs parses a string of args into a nice FluxConfig
+func ParseFluxArgs(argString string) (*FluxConfig, error) {
 	cfg := &FluxConfig{}
-	parser := flags.NewParser(cfg, flags.IgnoreUnknown)
-	_, err = parser.ParseArgs(strings.Split(out, " "))
-	if err != nil {
-		return nil, err
+
+	fs := pflag.NewFlagSet("default", pflag.ContinueOnError)
+	fs.ParseErrorsWhitelist.UnknownFlags = true
+
+	// strings
+	cfg.GitLabel = fs.String("git-label", "", "")
```

Comment:
If these (and the ff) were done with `fs.StringVar`, I think they would remain `nil` in the event that the flag isn't passed, obviating the cleanup code below.

Reply:
Oh, no, what am I saying. That's not how those work, they need a live pointer.